### PR TITLE
feat: duplicate selected/all saved objects UI

### PR DESCRIPTION
**Pull request by yubonluo**
### Description
duplicate selected/all saved objects UI
### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/server/saved_objects/routes/index.ts [hunk 1]
```diff
@@ -71,8 +71,8 @@ export function registerRoutes({
   registerLogLegacyImportRoute(router, logger);
   registerExportRoute(router, config);
   registerImportRoute(router, config);
-  registerResolveImportErrorsRoute(router, config);
   registerCopyRoute(router, config);
+  registerResolveImportErrorsRoute(router, config);
```
Comment (SuZhou-Joe) [inline src/core/server/saved_objects/routes/index.ts:75]: unnecessary change.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    };
+  }
+
+  workspaceToOption = (
+    workspace: WorkspaceAttribute,
+    currentWorkspaceName?: string
+  ): WorkspaceOption => {
+    // add (current) after current workspace name
+    let workspaceName = workspace.name;
+    if (workspace.name === currentWorkspaceName) {
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:104]: Better to use id to check if it's current workspace

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    const { duplicateMode } = this.props;
+    if (duplicateMode === DuplicateMode.All) {
+      const { allSelectedObjects } = this.state;
+      const categorizedObjects = groupBy(allSelectedObjects, (object) => object.type);
+      const savedObjectTypeInfoMap = new Map<string, [number, boolean]>();
+      for (const [savedObjectType, savedObjects] of Object.entries(categorizedObjects)) {
+        savedObjectTypeInfoMap.set(savedObjectType, [savedObjects.length, true]);
+      }
+      this.setState({ savedObjectTypeInfoMap });
+    }
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:144]: Seems the code should be move to constructor or beforeMount lifeCycle.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+    const targetWorkspace = this.state.targetWorkspaceOption[0].key;
+
+    await this.props.onDuplicate(
+      savedObjects,
+      this.state.isIncludeReferencesDeepChecked,
+      targetWorkspace!
+    );
+
+    if (this.isMounted) {
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:172]: If the `onDuplicate` gives error, we should keep the duplicate modal visible.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+      this.setState({
+        isLoading: false,
+      });
+    }
+  };
+
+  onSearchWorkspaceChange = (searchValue: string) => {
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:179]: I think ComboBox has default behvior of filtering by the search value, can we use that instead of filtering by ourself?

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+    const ignoredSelectedObjectsLength = selectedObjects.length - includedSelectedObjects.length;
+
+    let confirmDuplicateButtonEnabled = false;
+    if (!!targetWorkspaceId && includedSelectedObjects.length > 0) {
+      confirmDuplicateButtonEnabled = true;
+    }
+
+    const warningMessageForOnlyOneSavedObject = (
+      <p>
+        <b style={{ color: '#000' }}>1</b> saved object will <b style={{ color: '#000' }}>not</b> be
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:283]: https://oui.opensearch.org/1.5/#/display/text#coloring-text . Do not hard code color as we need to take dark mode into consideration.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+            label={
+              <FormattedMessage
+                id="savedObjectsManagement.objectsTable.duplicateModal.relatedObjects"
+                defaultMessage="Related Objects"
+              />
+            }
+          >
+            <>
+              <EuiText size="s" color="subdued">
+                {
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:373]: i18n

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx [hunk 8]
```diff
@@ -38,12 +38,60 @@ describe('Header', () => {
       onExportAll: () => {},
       onImport: () => {},
       onRefresh: () => {},
-      totalCount: 4,
+      onCopy: () => {},
+      onDuplicate: () => {},
+      title: 'Saved Objects',
+      selectedCount: 0,
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx:44]: Seems selectedCount is useless.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Represents the result of trying to duplicate the saved object.
+ * Contains `error` prop if something unexpected happened (e.g. network error).
+ * Contains an `id` if persisting was successful. If `id` and
+ * `error` are undefined, persisting was not successful, but the
+ * modal can still recover (e.g. the name of the saved object was already taken).
+ */
+
+export function showDuplicateModal(
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx:25]: This file is used for?

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx [hunk 10]
```diff
@@ -298,7 +303,7 @@ export class Table extends PureComponent<TableProps, TableState> {
           {
             name: i18n.translate(
               'savedObjectsManagement.objectsTable.table.columnActions.viewRelationshipsActionName',
-              { defaultMessage: 'Relationships' }
+              { defaultMessage: 'View object relationships' }
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:306]: Please keep the default message as it was because it may break cypress test.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 11]
```diff
@@ -671,6 +679,99 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
...
+        } else {
+          notifications.toasts.addDanger({
+            title: i18n.translate(
+              'savedObjectsManagement.objectsTable.duplicate.dangerNotification',
+              {
+                defaultMessage:
+                  'Unable to duplicate ' + savedObjects.length.toString() + ' saved objects',
+              }
+            ),
+          });
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:745]: Can we combine these two `else` into one?

File: src/plugins/workspace/public/workspace_client.ts [hunk 12]
```diff
@@ -112,7 +113,21 @@ export class WorkspaceClient {
...
+      const resultWithWritePermission = await this.list({
+        perPage: 999,
+        permissionModes: [WorkspacePermissionMode.LibraryWrite],
+      });
+      if (resultWithWritePermission?.success) {
+        const workspaceIdsWithWritePermission = resultWithWritePermission.result.workspaces.map(
+          (workspace: WorkspaceAttribute) => workspace.id
+        );
+        let workspaces = result.result.workspaces;
+        workspaces = result.result.workspaces.map((workspace: WorkspaceAttribute) => ({
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/workspace_client.ts:125]:
```suggestion
        const workspaces = result.result.workspaces.map((workspace: WorkspaceAttribute) => ({
```

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx (hunk 9 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx:25]: This file is used to reuse the duplicate logic in subsequent dashboard and visualize plugins.

File: src/core/server/saved_objects/routes/index.ts (hunk 1 above)
Comment (yubonluo) [reply src/core/server/saved_objects/routes/index.ts:75]: It has been modified

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 2 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:104]: Yes, it had been modified

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 3 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:144]: I will move the logic of obtaining workspace options to the `constructor()`. The logic of obtaining all saved objects will still be placed in `componentDidMount()`, otherwise the objects will not be obtained.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 4 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:172]: Yes, the logic has been implied in `src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx`.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 5 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:179]: Yes, I will delete the useless search logic.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 6 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:283]: Yes, I will modify the code.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 7 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:373]: Done

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx (hunk 10 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:306]: Done

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 11 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:745]: Yes, I will do this.

File: src/plugins/workspace/public/workspace_client.ts (hunk 12 above)
Comment (yubonluo) [reply src/plugins/workspace/public/workspace_client.ts:125]: Done

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx (hunk 8 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx:44]: Delete useless code

File: src/core/server/saved_objects/routes/copy.ts [hunk 13]
```diff
@@ -1,12 +1,19 @@
 /*
- * Copyright OpenSearch Contributors
  * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Any modifications Copyright OpenSearch Contributors. See
```
Comment (gaobinlong) [inline src/core/server/saved_objects/routes/copy.ts:8]: The code in copy.ts and copy.test.ts are old, please use the code in the `workspace-pr-integr` branch.

Comment (yubonluo) [reply src/core/server/saved_objects/routes/copy.ts]: Have reverted these codes

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx (hunk 9 above)
Comment (SuZhou-Joe) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx]: If we need to show a modal, we can embed it as a React component. Why do we make it as a util?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 14]
```diff
@@ -647,9 +657,7 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
       return null;
     }
     const { applications } = this.props;
-    const newIndexPatternUrl = applications.getUrlForApp('management', {
-      path: 'opensearch-dashboards/indexPatterns',
-    });
+    const newIndexPatternUrl = applications.getUrlForApp('indexPatterns');
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:660]:
```suggestion
    const newIndexPatternUrl = applications.getUrlForApp('management', {
      path: 'opensearch-dashboards/indexPatterns',
    });
```
Please do not change the index pattern url in this PR.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 15]
```diff
@@ -93,14 +94,15 @@ import {
   SavedObjectsManagementColumnServiceStart,
   SavedObjectsManagementNamespaceServiceStart,
 } from '../../services';
-import { Header, Table, Flyout, Relationships } from './components';
-import { DataPublicPluginStart } from '../../../../../plugins/data/public';
+import { Header, Table, Flyout, Relationships, SavedObjectsDuplicateModal } from './components';
+import { DataPublicPluginStart } from '../../../../data/public';
+
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:99]:
```suggestion
```

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 16]
```diff
@@ -93,14 +94,15 @@ import {
...
+import { DuplicateMode } from './';
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:100]: Should import from `./components` instead of `./index`. Import from `./index` makes the dependency chain confusing.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx [hunk 17]
```diff
@@ -233,6 +234,11 @@ describe('SavedObjectsTable', () => {
     }));
   });
 
+  afterEach(() => {
+    jest.clearAllMocks();
+    jest.resetModules();
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx:239]: I am thinking maybe we should only clear the needed mocks instead of clearAll.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx [hunk 18]
```diff
@@ -78,14 +80,15 @@ export interface TableProps {
   items: SavedObjectWithMetadata[];
   itemId: string | (() => string);
   totalItemCount: number;
-  onQueryChange: (query: any, filterFields: string[]) => void;
+  onQueryChange: (query: any, filterFields?: string[]) => void;
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:83]: Seems unrelated changes.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 19]
```diff
@@ -93,14 +94,15 @@ import {
   SavedObjectsManagementColumnServiceStart,
   SavedObjectsManagementNamespaceServiceStart,
 } from '../../services';
-import { Header, Table, Flyout, Relationships } from './components';
-import { DataPublicPluginStart } from '../../../../../plugins/data/public';
+import { Header, Table, Flyout, Relationships, SavedObjectsDuplicateModal } from './components';
+import { DataPublicPluginStart } from '../../../../data/public';
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:98]:
```suggestion
import { DataPublicPluginStart } from '../../../../plugins/data/public';
```

better to avoid unnecessary change.

File: src/plugins/saved_objects_management/public/lib/duplicate_saved_objects.ts [hunk 20]
```diff
@@ -0,0 +1,27 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Any modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/lib/duplicate_saved_objects.ts:10]: Done

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+      isIncludeReferencesDeepChecked,
+      allSelectedObjects,
+    } = this.state;
+    const { duplicateMode, onClose } = this.props;
+    const targetWorkspaceId = targetWorkspaceOption?.at(0)?.key;
+    let selectedObjects = allSelectedObjects;
+    if (duplicateMode === DuplicateMode.All) {
+      selectedObjects = selectedObjects.filter((item) => this.isSavedObjectTypeIncluded(item.type));
+    }
+    // If the target workspace is selected, all saved objects will be retained.
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:213]: Have modified it according to your suggestion.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 7 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:315]: Have modified it according to your suggestion.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx [hunk 22]
```diff
@@ -31,19 +31,59 @@
 import React from 'react';
 import { shallow } from 'enzyme';
 import { Header } from './header';
+import { any } from 'bluebird';
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx:34]: Have modified it according to your suggestion.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 14 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:660]: Have modified it according to your suggestion.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 15 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:99]: Have modified it according to your suggestion.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx (hunk 17 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx:239]: It seems that we don't any afterEach code.

Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx:239]: Have deleted the logic

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx (hunk 18 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:83]: Have deleted it according to your suggestion.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 19 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:98]: Have modified it according to your suggestion.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx (hunk 9 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/show_duplicate_modal.tsx]: The file has been deleted.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:1]: The code has been split to two files.
1. The enum file has been moved to types.ts;
2. The renderDuplicateObjectCategory and renderDuplicateObjectCategories function has been moved to a new tsx file.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx [hunk 23]
```diff
@@ -38,12 +38,51 @@ describe('Header', () => {
       onExportAll: () => {},
       onImport: () => {},
       onRefresh: () => {},
-      totalCount: 4,
+      onDuplicate: () => {},
+      title: 'Saved Objects',
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx:42]: title is not a required props for header I think.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+    if (duplicateMode === DuplicateMode.All) {
+      selectedObjects = selectedObjects.filter((item) => this.isSavedObjectTypeIncluded(item.type));
+    }
+    // If the target workspace is not selected, all saved objects will be retained.
+    // If the target workspace has been selected, filter out the saved objects that belongs to the workspace.
+    const includedSelectedObjects = selectedObjects.filter((item) =>
+      !!targetWorkspaceId && !!item.workspaces ? !item.workspaces.includes(targetWorkspaceId) : true
+    );
+
+    const ignoredSelectedObjectsLength = selectedObjects.length - includedSelectedObjects.length;
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:169]: Extract to a method of the component maybe? It is hard to understand what these lines are doing.

**Review by SuZhou-Joe — APPROVED** [review]
Looks much better than the first version, good job!

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx (hunk 24 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx]: Sure, have extracted the logic to a method.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx (hunk 23 above)
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/header.test.tsx]: Have deleted the useless code.

**Review by SuZhou-Joe — APPROVED** [review]
LGTM
